File: public/cpp/086-find-median-from-data-stream.cpp
/*
 * Problem: Find Median from Data Stream (LeetCode 295)
 * Link: https://leetcode.com/problems/find-median-from-data-stream/
 * Difficulty: Hard
 * Category: Heaps and Priority Queues
 * 
 * Description:
 * The median is the middle value in an ordered integer list. If the size of the
 * list is even, there is no middle value, and the median is the mean of the two
 * middle values.
 * 
 * Implement the MedianFinder class:
 * - MedianFinder() initializes the MedianFinder object.
 * - void addNum(int num) adds the integer num from the data stream to the data structure.
 * - double findMedian() returns the median of all elements so far.
 * 
 * Example 1:
 * Input:
 * ["MedianFinder", "addNum", "addNum", "findMedian", "addNum", "findMedian"]
 * [[], [1], [2], [], [3], []]
 * Output: [null, null, null, 1.5, null, 2.0]
 * 
 * Constraints:
 * - -10^5 <= num <= 10^5
 * - There will be at least one element in the data structure before calling findMedian.
 * - At most 5 * 10^4 calls will be made to addNum and findMedian.
 */

#include <iostream>
#include <queue>
#include <vector>
#include <algorithm>
#include <set>
using namespace std;

/*
 * APPROACH 1: TWO HEAPS (OPTIMAL)
 * 
 * Intuition:
 * - Use max heap for lower half of numbers
 * - Use min heap for upper half of numbers
 * - Keep heaps balanced (size difference <= 1)
 * - Median is at top of heap(s)
 * 
 * Algorithm:
 * addNum:
 * 1. Add to max heap (lower half)
 * 2. Move max heap top to min heap (upper half)
 * 3. Balance: if min heap larger, move top to max heap
 * 
 * findMedian:
 * 1. If heaps equal size: average of both tops
 * 2. Otherwise: top of larger heap
 * 
 * Time Complexity:
 * - addNum: O(log n) - heap operations
 * - findMedian: O(1) - just peek
 * Space Complexity: O(n) - store all numbers
 */
class MedianFinder1 {
private:
    priority_queue<int> maxHeap; // Lower half (max heap)
    priority_queue<int, vector<int>, greater<int>> minHeap; // Upper half (min heap)
    
public:
    MedianFinder1() {}
    
    void addNum(int num) {
        // Add to max heap first
        maxHeap.push(num);
        
        // Balance: move largest from lower half to upper half
        minHeap.push(maxHeap.top());
        maxHeap.pop();
        
        // Ensure max heap has equal or one more element
        if (minHeap.size() > maxHeap.size()) {
            maxHeap.push(minHeap.top());
            minHeap.pop();
        }
    }
    
    double findMedian() {
        if (maxHeap.size() > minHeap.size()) {
            return maxHeap.top();
        }
        return (maxHeap.top() + minHeap.top()) / 2.0;
    }
};

/*
 * APPROACH 2: SORTED VECTOR WITH BINARY SEARCH
 * 
 * Intuition:
 * - Maintain sorted array
 * - Use binary search to find insertion position
 * - Median is at middle position(s)
 * 
 * Algorithm:
 * addNum:
 * 1. Binary search for insertion position
 * 2. Insert at correct position
 * 
 * findMedian:
 * 1. Return middle element(s)
 * 
 * Time Complexity:
 * - addNum: O(n) - insertion in vector
 * - findMedian: O(1)
 * Space Complexity: O(n)
 */
class MedianFinder2 {
private:
    vector<int> nums;
    
public:
    MedianFinder2() {}
    
    void addNum(int num) {
        if (nums.empty()) {
            nums.push_back(num);
        } else {
            // Binary search for insertion position
            auto it = lower_bound(nums.begin(), nums.end(), num);
            nums.insert(it, num);
        }
    }
    
    double findMedian() {
        int n = nums.size();
        if (n % 2 == 1) {
            return nums[n / 2];
        }
        return (nums[n / 2 - 1] + nums[n / 2]) / 2.0;
    }
};

/*
 * APPROACH 3: MULTISET (BALANCED BST)
 * 
 * Intuition:
 * - Use multiset (self-balancing BST)
 * - Maintain iterator to median position
 * - Update iterator on insertions
 * 
 * Algorithm:
 * addNum:
 * 1. Insert into multiset
 * 2. Adjust median iterator
 * 
 * findMedian:
 * 1. Use median iterator
 * 
 * Time Complexity:
 * - addNum: O(log n) - BST insertion
 * - findMedian: O(1) - iterator access
 * Space Complexity: O(n)
 */
class MedianFinder3 {
private:
    multiset<int> data;
    multiset<int>::iterator lo, hi;
    
public:
    MedianFinder3() : lo(data.end()), hi(data.end()) {}
    
    void addNum(int num) {
        int n = data.size();
        data.insert(num);
        
        if (n == 0) {
            lo = hi = data.begin();
        } else if (n % 2 == 1) {
            // Odd to even: median is average of two middle
            if (num < *lo) {
                lo--;
            } else {
                hi++;
            }
        } else {
            // Even to odd: median is single middle
            if (num >= *lo && num < *hi) {
                lo++;
                hi--;
            } else if (num >= *hi) {
                lo++;
            } else {
                lo = --hi;
            }
        }
    }
    
    double findMedian() {
        return (*lo + *hi) / 2.0;
    }
};

/*
 * APPROACH 4: TWO HEAPS WITH SIZE TRACKING
 * 
 * Intuition:
 * - Similar to Approach 1 but with explicit size tracking
 * - Clearer logic for balancing
 * - Better for understanding
 * 
 * Algorithm:
 * Same as Approach 1 but with size variables
 * 
 * Time Complexity:
 * - addNum: O(log n)
 * - findMedian: O(1)
 * Space Complexity: O(n)
 */
class MedianFinder4 {
private:
    priority_queue<int> maxHeap;
    priority_queue<int, vector<int>, greater<int>> minHeap;
    int maxHeapSize, minHeapSize;
    
public:
    MedianFinder4() : maxHeapSize(0), minHeapSize(0) {}
    
    void addNum(int num) {
        if (maxHeapSize == 0 || num <= maxHeap.top()) {
            maxHeap.push(num);
            maxHeapSize++;
        } else {
            minHeap.push(num);
            minHeapSize++;
        }
        
        // Balance heaps
        if (maxHeapSize > minHeapSize + 1) {
            minHeap.push(maxHeap.top());
            maxHeap.pop();
            maxHeapSize--;
            minHeapSize++;
        } else if (minHeapSize > maxHeapSize) {
            maxHeap.push(minHeap.top());
            minHeap.pop();
            minHeapSize--;
            maxHeapSize++;
        }
    }
    
    double findMedian() {
        if (maxHeapSize > minHeapSize) {
            return maxHeap.top();
        }
        return (maxHeap.top() + minHeap.top()) / 2.0;
    }
};

/*
 * APPROACH 5: INSERTION SORT APPROACH
 * 
 * Intuition:
 * - Keep array sorted by inserting at correct position
 * - Simple but less efficient
 * - Good for small datasets
 * 
 * Algorithm:
 * addNum:
 * 1. Find position using linear search
 * 2. Insert at position
 * 
 * findMedian:
 * 1. Return middle element(s)
 * 
 * Time Complexity:
 * - addNum: O(n) - linear search + insertion
 * - findMedian: O(1)
 * Space Complexity: O(n)
 */
class MedianFinder5 {
private:
    vector<int> nums;
    
public:
    MedianFinder5() {}
    
    void addNum(int num) {
        int pos = 0;
        while (pos < nums.size() && nums[pos] < num) {
            pos++;
        }
        nums.insert(nums.begin() + pos, num);
    }
    
    double findMedian() {
        int n = nums.size();
        if (n % 2 == 1) {
            return nums[n / 2];
        }
        return (nums[n / 2 - 1] + nums[n / 2]) / 2.0;
    }
};

// Test function
void testMedianFinder(int approach) {
    cout << "Testing Approach " << approach << ":\n";
    
    if (approach == 1) {
        MedianFinder1 mf;
        cout << "addNum(1)\n";
        mf.addNum(1);
        cout << "addNum(2)\n";
        mf.addNum(2);
        cout << "findMedian() = " << mf.findMedian() << " (expected: 1.5)\n";
        cout << "addNum(3)\n";
        mf.addNum(3);
        cout << "findMedian() = " << mf.findMedian() << " (expected: 2)\n";
    } else if (approach == 2) {
        MedianFinder2 mf;
        cout << "addNum(1)\n";
        mf.addNum(1);
        cout << "addNum(2)\n";
        mf.addNum(2);
        cout << "findMedian() = " << mf.findMedian() << " (expected: 1.5)\n";
        cout << "addNum(3)\n";
        mf.addNum(3);
        cout << "findMedian() = " << mf.findMedian() << " (expected: 2)\n";
    } else if (approach == 3) {
        MedianFinder3 mf;
        cout << "addNum(1)\n";
        mf.addNum(1);
        cout << "addNum(2)\n";
        mf.addNum(2);
        cout << "findMedian() = " << mf.findMedian() << " (expected: 1.5)\n";
        cout << "addNum(3)\n";
        mf.addNum(3);
        cout << "findMedian() = " << mf.findMedian() << " (expected: 2)\n";
    } else if (approach == 4) {
        MedianFinder4 mf;
        cout << "addNum(1)\n";
        mf.addNum(1);
        cout << "addNum(2)\n";
        mf.addNum(2);
        cout << "findMedian() = " << mf.findMedian() << " (expected: 1.5)\n";
        cout << "addNum(3)\n";
        mf.addNum(3);
        cout << "findMedian() = " << mf.findMedian() << " (expected: 2)\n";
    } else if (approach == 5) {
        MedianFinder5 mf;
        cout << "addNum(1)\n";
        mf.addNum(1);
        cout << "addNum(2)\n";
        mf.addNum(2);
        cout << "findMedian() = " << mf.findMedian() << " (expected: 1.5)\n";
        cout << "addNum(3)\n";
        mf.addNum(3);
        cout << "findMedian() = " << mf.findMedian() << " (expected: 2)\n";
    }
    cout << "\n";
}

int main() {
    // Test Case 1: Basic operations
    cout << "Test Case 1: Basic operations\n";
    for (int i = 1; i <= 5; i++) {
        testMedianFinder(i);
    }
    
    // Test Case 2: More complex sequence
    cout << "Test Case 2: Complex sequence\n";
    MedianFinder1 mf;
    vector<int> nums = {5, 15, 1, 3, 8, 7, 9, 10, 20, 12};
    
    for (int num : nums) {
        mf.addNum(num);
        cout << "After adding " << num << ", median = " << mf.findMedian() << "\n";
    }
    
    // Test Case 3: Negative numbers
    cout << "\nTest Case 3: Negative numbers\n";
    MedianFinder1 mf2;
    vector<int> nums2 = {-1, -2, -3, -4, -5};
    
    for (int num : nums2) {
        mf2.addNum(num);
        cout << "After adding " << num << ", median = " << mf2.findMedian() << "\n";
    }
    
    return 0;
}

/*
 * COMPLEXITY ANALYSIS SUMMARY:
 * 
 * Approach 1 (Two Heaps - OPTIMAL):
 * - addNum: O(log n) - heap operations
 * - findMedian: O(1) - peek at tops
 * - Space: O(n)
 * - Best for: Most efficient, industry standard
 * 
 * Approach 2 (Sorted Vector):
 * - addNum: O(n) - insertion in vector
 * - findMedian: O(1)
 * - Space: O(n)
 * - Best for: Simple implementation
 * 
 * Approach 3 (Multiset):
 * - addNum: O(log n) - BST insertion
 * - findMedian: O(1) - iterator access
 * - Space: O(n)
 * - Best for: C++ specific solution
 * 
 * Approach 4 (Two Heaps with Tracking):
 * - addNum: O(log n)
 * - findMedian: O(1)
 * - Space: O(n)
 * - Best for: Clearer logic
 * 
 * Approach 5 (Insertion Sort):
 * - addNum: O(n) - linear search
 * - findMedian: O(1)
 * - Space: O(n)
 * - Best for: Small datasets
 * 
 * INTERVIEW TIPS:
 * 1. Start with Approach 1 (two heaps)
 * 2. Explain heap balancing strategy
 * 3. Draw diagram showing heap states
 * 4. Mention max heap stores smaller half
 * 5. Discuss why we keep heaps balanced
 * 
 * COMMON MISTAKES:
 * 1. Using min heap for lower half (wrong!)
 * 2. Not balancing heaps properly
 * 3. Integer overflow when averaging (use double)
 * 4. Wrong median calculation for even size
 * 5. Not handling empty data structure
 * 
 * FOLLOW-UP QUESTIONS:
 * 1. What if 99% of numbers are in [0,100]? (Use counting sort)
 * 2. Can you handle duplicates? (Yes, all approaches work)
 * 3. What if stream is infinite? (Two heaps still work)
 * 4. How to find kth percentile? (Adjust heap sizes)
 * 5. What if we need to remove numbers? (Use multiset)
 * 
 * RELATED PROBLEMS:
 * - Sliding Window Median
 * - Find K Pairs with Smallest Sums
 * - Kth Largest Element in Stream
 * - IPO (maximize capital)
 * - Meeting Rooms II
 * 
 * KEY INSIGHTS:
 * 1. Two heaps maintain sorted halves efficiently
 * 2. Max heap for lower half, min heap for upper half
 * 3. Keep heaps balanced (size diff <= 1)
 * 4. Median always at heap top(s)
 * 5. O(log n) insertion, O(1) median retrieval
 * 
 * HEAP BALANCING RULES:
 * 1. Max heap size = min heap size OR max heap size = min heap size + 1
 * 2. All elements in max heap <= all elements in min heap
 * 3. After each insertion, restore balance
 * 4. If odd total: median is max heap top
 * 5. If even total: median is average of both tops
 * 
 * WHY TWO HEAPS WORK:
 * - Max heap gives largest of smaller half in O(1)
 * - Min heap gives smallest of larger half in O(1)
 * - Together they give median in O(1)
 * - Insertion is O(log n) due to heap operations
 * - This is optimal for this problem
 */

// Made with Bob

File: src/data/tutorials/086-find-median-from-data-stream.js
/**
 * Tutorial content for #086 — Find Median from Data Stream
 * Keyed by approach name (must match the name parsed from the solution file).
 */
export const tutorial = {
  problem: {
    statement: `Design a data structure that supports two operations: addNum(int num) — add an integer from a stream, and findMedian() — return the median of all elements added so far. The median is the middle value for odd count, or the average of the two middle values for even count.`,
    example: `addNum(1) → stream: [1]         median = 1.0\naddNum(2) → stream: [1,2]       median = 1.5\naddNum(3) → stream: [1,2,3]     median = 2.0\n✅ Answer: 1.5 then 2.0`,
    keyInsight: `Keep the stream split into two halves: a max-heap for the lower half and a min-heap for the upper half. The median lives at the tops of these two heaps. Balance them after each insertion to keep their sizes differ by at most 1.`,
  },

  approaches: {
    'Two Heaps': {
      intuition: `Maintain two heaps:\n• lo: a max-heap of the smaller half of numbers.\n• hi: a min-heap of the larger half.\n\nInvariant: every element in lo ≤ every element in hi, and |lo.size - hi.size| ≤ 1.\n\nMedian = lo.top if lo is larger, hi.top if hi is larger, or (lo.top + hi.top) / 2.0 if equal sizes.`,
      steps: [
        `addNum(num):`,
        `  Push num into lo (max-heap).`,
        `  Move lo's max to hi (ensures lo.max ≤ hi.min always).`,
        `  If hi.size > lo.size: move hi's min back to lo (rebalance).`,
        `findMedian():`,
        `  If lo.size > hi.size: return lo.top.`,
        `  Return (lo.top + hi.top) / 2.0.`,
      ],
      example: `addNum(1): lo→push 1. move max(1) to hi. hi:[1], lo:[]. hi>lo → move 1 back. lo:[1], hi:[]\naddNum(2): lo→push 2. move max(2) to hi. lo:[1], hi:[2]. equal size.\nfindMedian → (1+2)/2 = 1.5 ✅\naddNum(3): lo→push 3. move max(3) to hi. lo:[1], hi:[2,3]. hi>lo → move min(2) to lo. lo:[2,1], hi:[3]\nfindMedian → lo.size>hi.size → lo.top = 2.0 ✅`,
      keyInsight: `O(log n) per addNum (heap insertion), O(1) per findMedian (peek tops). O(n) space. This is the canonical and optimal solution to this problem.`,
    },

    'Sorted Array with Binary Search': {
      intuition: `Maintain a sorted array. To add a number, binary search for its insertion position and insert it (shifting elements). findMedian just reads the middle index.`,
      steps: [
        `Maintain a sorted list.`,
        `addNum(num): use binary search to find insertion point; insert.`,
        `findMedian(): n = list.size. If n is odd return list[n/2]. If even return (list[n/2-1] + list[n/2]) / 2.0.`,
      ],
      example: `addNum(1): [1]\naddNum(2): binary search → insert at 1 → [1,2]. median=(1+2)/2=1.5\naddNum(3): insert at 2 → [1,2,3]. median=list[1]=2.0 ✅`,
      keyInsight: `O(n) per addNum (insertion shift), O(1) per findMedian. O(n) space. Fine for small streams but too slow for large inputs — the Two Heaps approach is strictly better.`,
    },

    'Two Heaps with Explicit Routing': {
      intuition: `Same as Two Heaps, but with an explicit rule for where each new number goes: if num ≤ lo.top push to lo, else push to hi. Then rebalance by size. This makes the routing decision visible and avoids the "push then pop" pattern.`,
      steps: [
        `addNum(num):`,
        `  If lo is empty OR num ≤ lo.top: push num to lo.`,
        `  Else: push num to hi.`,
        `  Rebalance: while |lo.size - hi.size| > 1, move top of larger to smaller.`,
        `findMedian(): same as Two Heaps.`,
      ],
      example: `addNum(5): lo empty → push to lo. lo:[5], hi:[]\naddNum(3): 3≤5 → push to lo. lo:[5,3]. |lo|-|hi|=2>1 → move 5 to hi. lo:[3], hi:[5]\naddNum(4): 4≤5 (lo.top=3? no 4>3) → push to hi. hi:[4,5]. |hi|>|lo| → move 4 to lo. lo:[4,3], hi:[5]\nfindMedian → (4+5)/2 = 4.5 ✅`,
      keyInsight: `O(log n) per addNum, O(1) per findMedian. Functionally identical to the basic Two Heaps — this variant just makes the routing logic explicit, which some find easier to reason about.`,
    },

    'Insertion Sort (Linear Insert)': {
      intuition: `Keep a plain list and maintain it in sorted order by inserting each new number at the correct position (linear scan to find it). findMedian reads the center. Simple but slow — practical only for small data.`,
      steps: [
        `Maintain List<Integer> sorted.`,
        `addNum(num): scan from the start to find where num fits; insert there.`,
        `findMedian(): read from middle index.`,
      ],
      example: `addNum(3): [3]\naddNum(1): scan: 1<3 → insert at 0 → [1,3]. median=(1+3)/2=2.0\naddNum(2): scan: 2<3 → insert at 1 → [1,2,3]. median=2.0 ✅`,
      keyInsight: `O(n) per addNum, O(1) per findMedian. O(n) space. The simplest correct implementation. Only use for tiny inputs or as a reference baseline.`,
    },

    'TreeMap (Balanced BST)': {
      intuition: `Use a TreeMap (Red-Black tree) to maintain elements in sorted order. Track the total count and a "lower" pointer to the element at position count/2. addNum keeps this pointer updated by shifting it when needed.`,
      steps: [
        `TreeMap<Integer,Integer> map (value → frequency).`,
        `Track lo (the "lower median" element) and loCount (how many of lo are in the lower half).`,
        `addNum(num): insert into map. Adjust lo and loCount based on where num falls relative to lo.`,
        `findMedian(): if even, (lo + map.higherKey(lo))/2.0. If odd, return lo.`,
      ],
      example: `addNum(1): map:{1:1}, lo=1, total=1. median=1.0\naddNum(2): map:{1:1,2:1}, total=2. lo stays 1.\nfindMedian → even: (1 + nextKey(1)=2)/2 = 1.5 ✅\naddNum(3): map:{1:1,2:1,3:1}, total=3. lo moves to 2.\nfindMedian → odd: lo = 2.0 ✅`,
      keyInsight: `O(log n) per addNum and findMedian. O(n) space. Handles duplicate values elegantly and is useful when you also need range queries — but the Two Heaps approach is simpler for pure median finding.`,
    },
  },
}
